pgmax/factors/logical.py

```diff
@@ -0,0 +1,131 @@
+"""Defines an enumeration factor"""
+
+from dataclasses import dataclass
+from typing import Mapping, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+from pgmax import utils
+from pgmax.fg import nodes
+
+ALLOWED_LOGICAL_TYPES = ["OR"]
+
+
+@jax.tree_util.register_pytree_node_class
+@dataclass(frozen=True, eq=False)
+class LogicalWiring(nodes.Wiring):
+    """Wiring for LogicalFactors.
+
+    Args:
+        edges_num_states: Array of shape (num_edges,)
+            Number of states for the variables connected to each edge
+        var_states_for_edges: Array of shape (num_edge_states,)
+            Global variable state indices for each edge state
+        parents_edge_states: Array of shape (num_parents, 2)
+            parents_edge_states[ii, 0] contains the global factor index,
+            which takes into account all the LogicalFactors of the same subtype (AND / OR).
+            parents_edge_states[ii, 1] contains the message index of the parent variable's state 0,
+            which takes into account all the EnumerationFactors and LogicalFactors.
+            The parent variable's state 1 is parents_edge_states[ii, 2] + 1.
+        children_edge_states: Array of shape (num_factors,)
+            children_edge_states[ii] contains the message index of the child variable's state 0,
+            which takes into account all the EnumerationFactors and LogicalFactors.
+            The child variable's state 1 is children_edge_states[ii, 1] + 1.
+    """
+
+    edges_num_states: Union[np.ndarray, jnp.ndarray]
+    var_states_for_edges: Union[np.ndarray, jnp.ndarray]
+    parents_edge_states: Union[np.ndarray, jnp.ndarray]
+    children_edge_states: Union[np.ndarray, jnp.ndarray]
+
+
+@dataclass(frozen=True, eq=False)
+class LogicalFactor(nodes.Factor):
+    """A logical OR/AND factor
+    Args:
+        variables: List of connected variables.
+            The last variable is assumed to be the child one.
+        logical_type: The logical condition supported by the factor
+
+    Raises:
+        ValueError: If:
+            (1) The variables are not all binary
+            (2) There are less than 2 variables
+    """
+
+    variables: Tuple[nodes.Variable, ...]
+    logical_type: str
+    log_potentials: None
```

Comment:
Having this here means users can actually overwrite it with a non-None value. Replacing with `field(init=False, default=None)` would be better.